R/bioc-standalone.R

```diff
@@ -0,0 +1,217 @@
+    if (forget || is.null(version_map)) {
+      txt <- get_yaml_config(forget)
+      grps <- grep("^[^[:blank:]]", txt)
+      start <- match(grep("r_ver_for_bioc_ver", txt), grps)
+      map <- txt[seq(grps[start] + 1, grps[start + 1] - 1)]
+      map <- trimws(gsub("\"", "", sub(" #.*", "", map)))
+      pattern <- "(.*): (.*)"
+      bioc <- package_version(sub(pattern, "\\1", map))
+      r <- package_version(sub(pattern, "\\2", map))
+      status <- rep("out-of-date", length(bioc))
+      release <- get_release_version()
+      devel <- get_devel_version()
+      status[bioc == release] <- "release"
+      status[bioc == devel] <- "devel"
+
+      # append final version for 'devel' R
+      bioc <- c(
+        bioc, max(bioc)
+        # package_version(paste(unlist(max(bioc)) + 0:1, collapse = "."))
+      )
+      r <- c(r, package_version(paste(unlist(max(r)) + 0:1, collapse = ".")))
+      status <- c(status, "future")
+
+      version_map <<- rbind(
+        .VERSION_MAP_SENTINEL,
+        data.frame(
+          bioc_version = bioc, r_version = r,
+          bioc_status = factor(
+            status,
+            levels = c("out-of-date", "release", "devel", "future")
+          )
+        )
+      )
+    }
+    version_map
+  }
+
+  get_matching_bioc_version <- function(r_version = getRversion(),
+                                        forget = FALSE) {
+
+    minor <- get_minor_r_version(r_version)
+    if (minor %in% names(builtin_map)) return(builtin_map[[minor]])
+
+    # If we are not in the map, then we need to look this up in
+    # YAML data.
+
+    map <- get_version_map(forget = forget)
+    mine <- match(package_version(minor), map$r_version)
+    if (!is.na(mine)) return(map$bioc_version[mine])
+
+    # If it is not even in the YAML, then it must be some very old
+    # or very new version. If old, we fail. If new, we assume bioc-devel.
+    if (package_version(minor) < "2.1") {
+      stop("R version too old, cannot run Bioconductor")
+    }
+
+    get_devel_version()
+  }
+
+  get_bioc_version <- function(r_version = getRversion(),
+                               forget = FALSE) {
+    if (nzchar(v <- Sys.getenv("R_BIOC_VERSION", ""))) {
+      return(package_version(v))
+    }
+    get_matching_bioc_version(r_version, forget = forget)
+  }
+
+  get_repos <- function(bioc_version = "auto", forget = FALSE) {
+    if (identical(bioc_version, "auto")) {
+      bioc_version <- get_bioc_version(getRversion(), forget)
+    } else {
+      bioc_version <- package_version(bioc_version)
+    }
+    mirror <- Sys.getenv("R_BIOC_MIRROR", "https://bioconductor.org")
+    mirror <- getOption("BioC_mirror", mirror)
+    repos <- c(
+      BioCsoft      = "{mirror}/packages/{bv}/bioc",
+      BioCann       = "{mirror}/packages/{bv}/data/annotation",
+      BioCexp       = "{mirror}/packages/{bv}/data/experiment",
+      BioCworkflows =
+        if (bioc_version >= "3.7") "{mirror}/packages/{bv}/workflows",
+      BioCextra     =
```

Comment:
Is there any way for us to detect if the set of BioConductor repositories changes in new releases? Or will we have to update this list as chose changes happen?

Reply:
Good question! Unfortunately, AFAICT we can't easily automate this. The yaml file has the repos for the current devel version:
```yaml
## CHANGE THIS (i.e., uncomment) as various parts of the new devel version 
## become available. set to "[]" if none are available.
devel_repos:
- "bioc"
- "data/experiment"
- "data/annotation"
- "workflows"
```
but this is all, no repos for older versions, etc.

BiocManager has them hard-coded: https://github.com/Bioconductor/BiocManager/blob/6f8252609ddb9d2e02efbff96267ffb195ce1f32/R/repositories.R#L35-L36

There is also `tools:::.get_standard_repository_URLs()`, but that does not even have the correct repos now, and it is not updated often enough, anyway.
```r
❯ tools:::.get_standard_repository_URLs()
                                                   CRAN
                           "https://CRAN.R-project.org"
                                               BioCsoft
           "https://bioconductor.org/packages/3.9/bioc"
                                                BioCann
"https://bioconductor.org/packages/3.9/data/annotation"
                                                BioCexp
"https://bioconductor.org/packages/3.9/data/experiment"
```

Is there anything else I am missing?

What we could do is adding a test case that checks that (for R-devel/Bioc-devel?) we do have the correct set of repos, using the BiocManager package, which we would suggest. Then at least we will notice early if the repos change.

Reply:
I don't think there is anywhere else and in general the types of repos changes quite infrequently, so it shouldn't be too much work to keep track of manually.